### DISCUSSION TITLE
How to auto-hide modules when the data source is not available (e.g. battery or backlight)?

### DISCUSSION THREAD

**Discussion by unclechu** [polybar/polybar#3079] [Q&A]
Is it possible to auto-hide some modules such as battery or backlight when there is either no battery or backlight on a device? For instance if I just turn off the builtin display on a laptop (in favor of external one(s)) instead of “backlight” level I get a string like `%percentage%%` on the bar. I would like to just hide the whole module if the data is not available.

**Comment by unclechu** [comment]
I think I figured out why this happens by looking at the code. It could actually be considered a bug I think.

In the `include/modules/backlight.hpp` the `m_percentage` is initialized with 0:

``` c++
int m_percentage = 0;
```

By default “backlight” module reads brightness level from `actual_brightness` file. When the embedded display is turned off that file contains `0` in it:

``` console
$ cat /sys/class/backlight/intel_backlight/actual_brightness
0
```

In `src/modules/backlight.cpp` file in `bool backlight_module::on_event(const inotify_event& event)` function there is this chunk of code:

``` c++
    m_max_brightness = m_max.read();
    float val = m_val.read();
    int percentage = math_util::percentage(val, m_max_brightness);

    if (m_percentage != percentage) {
      // ...
      m_percentage = percentage;
      if (m_label) {
        m_label->reset_tokens();
        m_label->replace_token("%percentage%", to_string(m_percentage));
      }
      // ...
    }
```

So from what I can tell it replaces the token only if the value changed since the last check. But at the beginning initial *previous* value (`m_percentage`) is equal to `percentage` because `actual_brightness` is `0`. Thus the `%percentage%` token never gets replaced. That’s why I see `%percentage%` as text on the bar.

So basically if initial brightness is 0 you have this bug. Since `m_percentage` is just `int` I think the fix it to initialize it with `-1` since real brightness value is never below 0. Thus on initial check the token always get replaced.

**Reply by unclechu** [reply]
I’ll make a merge request, this post is just about the bug. But my original question is still open.

**Reply by unclechu** [reply]
The bugfix: https://github.com/polybar/polybar/pull/3080

**Comment by patrick96** [comment]
Coming back to the original question.

Both the battery and backlight module will refuse to start if there it can't find the configured backlight or battery.

For the backlight, I'm not sure the situation you describe is distinguishable from the user simply turning their screen brightness all the way to 0. The backlight and all the backing files still exist, it just shows 0 for the brightness.
